pandas/io/parsers.py

```diff
@@ -997,14 +997,14 @@ def _validate_usecols_arg(usecols):
 
     if usecols is not None:
         if callable(usecols):
-            return usecols
+            return usecols, None
         usecols_dtype = lib.infer_dtype(usecols)
         if usecols_dtype not in ('empty', 'integer',
                                  'string', 'unicode'):
             raise ValueError(msg)
 
-        return set(usecols)
-    return usecols
+        return set(usecols), usecols_dtype
```

Comment:
Early conversion to `set` has been used for quite some time, especially with the C engine.  Also, a `list` is not necessarily going to help with ordering I might add:
~~~python
>>> from pandas import read_csv
>>> from pandas.compat import StringIO
>>>
>>> data = 'a,b,c\n1,2,3'
>>> read_csv(StringIO(data), usecols=[2, 1])
   b  c
0  2  3
~~~
When I say "order," it is not about maintaining the order in which `usecols` is passed in but about ordering `usecols` correctly so that the names or indices are in the same order as they are listed in the data columns.  That is why converting at the last moment does not present any benefit.